### PR TITLE
A New Dawn :tm:

### PR DESCRIPTION
**Pull request by ShiftyRail**
For real this time.
Maybe.

I know what to try next. Using this for test-serb purposes.

As of late 2023, screenshots and videos : 

image
 
image

image

image

apc glow

shuttlecalled

:cl:

- experimental: A new lighting system is being tested on live. Please direct all feedback and comments to the github.

### CODE REVIEW

**Comment by ghost** [comment]
prepare for unforeseen consequences
genuinely

**Comment by Eneocho** [comment]
https://www.youtube.com/watch?v=xtrqYdvZ29E&t=110s

**Comment by Hinaichigo** [comment]
The performance is pretty smooth. Two things I noticed testing it out is that there are some erratic shadows, pictured below, and also, when you hold a lamp and walk into a dark area, the emitted light will jump a little bit in front of you and snap to the nearest tile, while your sprite will move smoothly.

europa21
europa1

**Comment by Exxion** [comment]
I don't remember those wacky shadows from the previous tests. I mean they still had wacky shadows, but not exactly those ones

**Comment by Hinaichigo** [comment]
Is it possible that it can vary with drivers/OS?

**Comment by Exxion** [comment]
I doubt it, but maybe. But he probably just did more work on it.

**Comment by ShiftyRail** [comment]
I don't get wacky shadows while testing on my end, so it might be an emerging effect after some time passes and some unknown interactions occur.

image

This is how this area looks like on my end.

**Comment by ShiftyRail** [comment]
@Hinaichigo out of curiosity, what's your BYOND version? Shadows relies a lot on `render_source` things so it might be why it's different client to client.

**Comment by Hinaichigo** [comment]
Version 514. I'm not sure how to view the minor version number.
It's running in a w10 VM hosted on linux. I disabled 3D acceleration in the virtualbox settings and it fixed the shadows, but it also made the performance a lot worse, similar to the previous time it was up on the test server.

**Comment by ShiftyRail** [comment]
>Linux

might explain it
Try the latest minor just to see

**Comment by Hinaichigo** [comment]
Updating to the latest minor, with 3D acceleration turned on, it has somewhat choppy performance and the shadow glitch remains.
I wonder if [this issue](http://www.byond.com/forum/post/2662168) is related. It's recent, so maybe it'll be fixed in an upcoming version.

**Comment by Exxion** [comment]
Playing on Linux through *Wine*, I don't see any of those weird shadows, though I do notice rather shit performance, particularly around the bridge. I notice that the client profiler says that basically all of the time taken is in one particular step. I wonder if it's possible to optimize it somewhat. Not necessarily a large concern if the performance issues aren't on Windows, though.

**Comment by Hinaichigo** [comment]
Finally got Wine set-up and it actually runs a lot more smoothly, maybe a little bit of slowdown near the brig. And the glitched shadows seem to be fixed.

The border of the table (where the glass is) still seems a little off somehow:
europa3

The issue with the flashlight light jumping out ahead also seems somewhat less stark at proper framerates.
Also, it seems like "disable flashlight" is broken on the PDA: if you disable the flashlight it doesn't turn off until you put it into your backpack.

Overall it looks really nice aesthetically.

**Comment by ghost** [comment]
> The performance is pretty smooth.

yeah about that... this still crawls to 8fps (about 15 flares per tile)
image

**Comment by Exxion** [comment]
Oh no way it gets laggy if someone makes a lagmachine

**Comment by adacovsk** [comment]
Not to be a critic, I like the lighting. There's several particles that still need lighting though, like fireball, and a bunch of others. The roulette revolver provides a wide array of various ones that need lighting.

Though perhaps objects like flares should be handled differently to avoid lag.

Not sure if this was tested, but turning on all lights seems to make it a little sluggish? Maybe it's just me

**Comment by Hinaichigo** [comment]
> I notice that the client profiler says that basically all of the time taken is in one particular step. I wonder if it's possible to optimize it somewhat.

What step is that? Also, how do you view that?

**Comment by Exxion** [comment]
>What step is that? Also, how do you view that?

Don't remember the name exactly. Something about grouping and setting bounds, I think.
If you're playing under Wine, you probably need admin abilities to do it, because you need to be able to right click the title bar and select an option there to do it normally. With admin, you can select profiler under the server dropdown at the top left, and click the client tab on that window.

**Comment by Exxion** [comment]
Anyway, as for client lag, moving lights shouldn't be significantly laggier than stationary lights. They're much heavier on the server, though.

**Comment by ShiftyRail** [comment]
>There's several particles that still need lighting though, like fireball, and a bunch of others. The roulette revolver provides a wide array of various ones that need lighting.

One step at a time.

**Comment by ShiftyRail** [comment]
Fixed an issue someone reported with cult buildings, + an attempt at making lights move smoothly

**Comment by Cartlord** [comment]
What's the actual aim of this P.R.? You didn't describe it at all. What does it add?

**Comment by Eneocho** [comment]
> What's the actual aim of this P.R.? You didn't describe it at all. What does it add?

The latest attempt to make EuropaLights work. See #31641, #31620, #30497 and #29551.

**Comment by Hinaichigo** [comment]
After testing out Wine for a little bit, there are also various issues (not with lighting, but in general) that aren't present when running in a VM, though using Wine is smoother performance wise. I know Linux isn't officially supported, but since it works pretty well now, and a few players, in addition to myself, use a Windows VM from what I understand, it would be nice if breaking compatibility could be avoided, in terms of both performance and the lighting system.

Edit: After using Wine a bit more it is way smoother and things look nicer. Although there are still some issues here and there.

**Review by adacovsk — COMMENTED** [review]
racial_powers.dm is now in another file called species_powers.dm, also I changed around invisibility and alphas for mobs and objects, but your `alpha_obj` makes sense for obj although it's `body_alpha` for carbons and there's still `alpha` for mobs too. You could probably separate out all the vision stuff for a great PR that might fix a lot of vision issues

**Comment by Eneocho** [comment]
Any updates on this one @ShiftyRail? Is the dream of fancy lights dead?

**Comment by adacovsk** [comment]
Should probably close this, as hopefully my darkness and planes PR will be merged and would result in horrible conflicts. Also the test server doesn't even work so you can't even test the lights anymore.

File: code/modules/lighting/light_atom_set.dm [hunk 1]
```diff
@@ -0,0 +1,113 @@
+// Destroys and removes a light
+/atom/proc/kill_light()
+	if(light_obj)
+		qdel(light_obj)
+		light_obj = null
+	if (shadow_obj)
+		qdel(shadow_obj)
+		shadow_obj = null
+	if (smooth_light_obj)
```
Comment (adacovsk) [inline code/modules/lighting/light_atom_set.dm]: shouldn't it be qdel(smooth_light_obj)

Comment (Exxion) [reply code/modules/lighting/light_atom_set.dm]: It should be and `QDEL_NULL()` is one of TG's actually reasonable macros and we should steal it

Comment (SECBATON-GRIFFON) [reply code/modules/lighting/light_atom_set.dm]: we just did. it's time

**Comment by SonixApache** [comment]
when

**Comment by ShiftyRail** [comment]
holy fuck, a year?

**Comment by Inorien** [comment]
> holy fuck, a year?

is it ready yet

**Comment by xeroxemnas** [comment]
Quality shiftycode

**Comment by Eneocho** [comment]
2 years since the new dawn was promised.

**Comment by Eneocho** [comment]
oh shit it lives

**Comment by Hinaichigo** [comment]
Please fix the icon diff some time so we can test it out.

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
does this support adding multiple "naked" datum/light_source to a single atom? that could be useful

**Comment by ShiftyRail** [comment]
This is finished, and probably as good as I'll ever get it.

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
something seems glitched with the walls here (on the test server):

https://github.com/vgstation-coders/vgstation13/assets/149450117/4a4a25e6-b0e6-43c7-bd65-8a69332df91d


and the hard shadow line seems a bit too stark here,
and there is that triangular shadow on the door:
image

it seems like there are little sub-tilesize shadows that look starker than they should here and there

i wonder if increasing the diffusion would look nicer?

i wonder how grue vision looks

**Comment by ShiftyRail** [comment]
@xXxEMBERGHOSTSQUADDIExXx Yeah as for the walls I'm not sure if there's anything else to do tbh. I've already did the best on that front. Worth noting that goonlights walls also look fucky.

Here's a fixed lag machine:


https://github.com/vgstation-coders/vgstation13/assets/31417754/73de83a0-485d-4258-9d3a-f97de43989a2

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
what's the reason the wall shading snaps depending on player position like that?

**Comment by ShiftyRail** [comment]
@xXxEMBERGHOSTSQUADDIExXx They are set to dynamically render depending whether or not you see the light. They do not render if there isn't a clear path between you and the light tile.

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
so maybe something is bugging with regard to detecting whether you can see the light

I think as is most players wouldnt find it acceptable with shadows changing just by moving from tile to tile

But since we can describe it verbally here I bet there's a way to fix it code wise no?

considering this a bit,
basically every "opaque" atom eg. wall needs to have a list of all the light sources impinging on it, and which sides of the wall each light is impinging on

then the client needs to only render the light components where that side of the wall is visible to the player

something like this conceptually:

```
/atom
	var/list/impinging_light_sources = list()

/atom/proc/add_impinging_light_source(var/datum/light_source/L)
	var/impinged_faces = get_impinged_faces(L) //returns NORTH, SOUTH, (NORTH | EAST), etc.
	impinging_light_sources[L] = impinged_faces
	
/mob/proc/render_lights_on_wall(var/turf/illuminated_wall)
	for(var/datum/light_source/L in illuminated_wall.impinging_light_sources)
		if(src.proc_that_returns_wall_faces_cardinal_directions_visible_by_mob(illuminated_wall) & illuminated_wall.impinging_light_sources[L])
			src.render_light_source_illumination_component(illuminated_wall, L)
```

**Comment by ShiftyRail** [comment]
@xXxEMBERGHOSTSQUADDIExXx Commendable effort, but the reality is that we don't really have that much flexibility in how images are sent to clients.

Unless I'm not reading the ref properly, there are two majors ways to send images:

- to one (1) client in particular
- to everyone (they're part of the world)

Sending images on a client-by-client basis is very taxing (this is how one-way windows work, and they're shit). The way it works rn, I'm using the light source itself as view-independent, but the wall lighting is view-dependent, so that you don't have walls lighting up in the middle of maint.

This view-dependency seems to cause some of the problems in rendering. Surprisingly, just making the wall shadows view-independant doesn't fully fix it - probably because of some intricacies with layers, different masks, etc. It just make it so you can see some walls being lit from outside the room without fixing the jumpyness (arguably, makes it worse).

One solution I did find (vid attached for a demo) is to just slightly shift the wall shadow rendering point to be more in the center of the hallways rather than sticking to walls. It doesn't 100% get rid of it, but it does make it perform significantly better.

https://github.com/vgstation-coders/vgstation13/assets/31417754/ddfac482-d946-4add-96c7-2a679a59068a

**Comment by ShiftyRail** [comment]
Additionally, an image sent to a client wouldn't work unless it's the full light plane that's sent -- all at once --. So we would have one light plane per client... that'd probably be too much.

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
What is the mechanism that allows vampire thrall indicators to only show for the players involved? Maybe you could compute a shadow image that's the same size as the players view range maybe with a one tile buffer zone depending on how movement is handled and then overlay that over the players view

**Comment by ShiftyRail** [comment]
Thrall indicators are on a universal plane we allow vampire players to have
access to.

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
One plane per player might not be so bad

**Comment by ShiftyRail** [comment]
I dont we can do much here because the access we have to client-side rendering is quite limited. The only feature in game that actually uses it proemintently are one-way windows and they lag a lot.

It's incredibly involved to make light sources check periodically for clients in their view and then compose the image, and then send it. And I doubt it would make much difference: I checked if the jumpyness changes significantly if you remove the KEEP_TOGETHER flag, and it doesn't, because BYOND is still doing some vision checks that I have no idea about.

As shown in the latest commit, the best way to address this is to slightly shift the rendering point of the lights. It improved it quite a bit

**Comment by ghost** [comment]
> Here's a """"fixed"""" """"lag machine"""" aka 15 flares visibly causing the game to run in 5 fps

translights4
translights3
translights2
translights1

maxresdefault

**Comment by ShiftyRail** [comment]
Thanks for your input Lulza. I look forward to your input when this is test-merged and hopefully you will change your mind when you see them live.

What am I saying? Of course you won't, because you're already permabanned. Lmao.

**Comment by ghost** [comment]
Sorry ShiftyFail, but that would imply you've managed to convince anyone in the slightest that this fugly client-heavy half-a$sed unfinished laggy pile of sh1tcode bloat and wasted human hours is somehow better than the normal old goonlights.

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
good that the jumpiness was seemingly fixed
I'll try it it out in a bit
I'd say the most glaring things aside from that I noticed so far are:

- the sudden shadow differences between adjacent walls.
- the 'drop shadow' look on the borders of many walls and floor tiles where a little section of floor right on the perimeter of the wall is darker than the rest of the floor

Also with the light bulb being on top of the shadows how does that work? Every light source or only certain ones? I notice the round light bulbs look the same in the before and after pics

**Comment by ShiftyRail** [comment]
>the sudden shadow differences between adjacent walls.

That's because I want the light level on the walls to decrease as the light effect is further away. This could be slightly improved but I want to get the bulk of Europa merged and then I will be able to experiment with the wall shadows a bit better. Serialize the objectives.

>drop shadow

This bit is mostly "accidental" and a result of post-processing. I heard mixed things about it. Some people like it, other dont. I think it's done more elegantly than how TG does it, though.

As for the light fixtures, this is on an atom by atom basis. You need to draw a moody light spirite for each atom. At the moment I did something for tubes and bulbs, but I think I messed up the dirs in the bulb sprite.

**Comment by ShiftyRail** [comment]
Issues as reported by the threads:

OH GOD IT'S LOOSE:

Ghost performance. Have an idea how to address it but it's ugly code. It should work in theory though.

To fix (bad):

Moody lights being fucked, especially bulbs and beam -- Easy
Incorrect shadows being projected on walls for pda sources -- gonna be medium to hard
Flirts, pda lights, and some light sources being killed when put in the backbag -- easy-ish

Doors opacity being fucked at map load -- probably just a matter of initialize()
Vox outpost looks really bad -- maybe hard

Non-bugs:

Sparks are too bright -- will be adjusted in a seperate PR post release

**Pull request closed without merging**